ide/csl.api/src/org/netbeans/modules/csl/editor/semantic/GsfSemanticLayer.java

```diff
@@ -53,19 +50,17 @@
  * 
  * @author Tor Norbye
  */
-public class GsfSemanticLayer extends AbstractHighlightsContainer implements DocumentListener {
+public final class GsfSemanticLayer extends AbstractHighlightsContainer implements DocumentListener {
     
     // -J-Dorg.netbeans.modules.csl.editor.semantic.GsfSemanticLayer.level=FINE
     private static final Logger LOG = Logger.getLogger(GsfSemanticLayer.class.getName());
 
-    //private Map<Token, Coloring> colorings;
-    private SortedSet<SequenceElement> colorings;
-    private int version;
+    private List<SequenceElement> colorings = List.of();
     private List<Edit> edits;
-    private Map<Language,Map<Coloring, AttributeSet>> CACHE = new HashMap<Language,Map<Coloring, AttributeSet>>();
-    private Document doc;
-    private List<Lookup.Result> coloringResults = new ArrayList<Lookup.Result>(3);
-    private List<LookupListener> coloringListeners = new ArrayList<LookupListener>(3);
+    private final Map<Language,Map<Coloring, AttributeSet>> cache = new HashMap<>();
+    private final Document doc;
+    private final List<Lookup.Result> coloringResults = new ArrayList<>(3);
+    private final List<LookupListener> coloringListeners = new ArrayList<>(3);
```

Comment:
My take on the reason for this:

```suggestion
    // Write only - used to keep strong reference to Lookup.Result/LookupListener
    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
    private final List<Lookup.Result> coloringResults = new ArrayList<>(3);
    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
    private final List<LookupListener> coloringListeners = new ArrayList<>(3);
```